patches/squirrel.mac/feat_add_new_squirrel_mac_bundle_installation_method_behind_flag.patch

```diff
@@ -0,0 +1,128 @@
+non-admin users to update applications when appropriately granted group based permissions.
+
+E.g. 775 and chown :staff
+
+diff --git a/Squirrel/SQRLInstaller.m b/Squirrel/SQRLInstaller.m
+index 7dd98ddee4ae0f4e01fd7aaa3486083bff7d0da1..c1f328fa8c3689218ef260347cb8f9d30b789efe 100644
+--- a/Squirrel/SQRLInstaller.m
++++ b/Squirrel/SQRLInstaller.m
+@@ -249,6 +249,7 @@ - (RACSignal *)acquireTargetBundleURLForRequest:(SQRLShipItRequest *)request {
+ 		] reduce:^(NSURL *directoryURL, SQRLCodeSignature *codeSignature) {
+ 			NSURL *targetBundleURL = request.targetBundleURL;
+ 			NSURL *newBundleURL = [directoryURL URLByAppendingPathComponent:targetBundleURL.lastPathComponent];
++			[NSFileManager.defaultManager createDirectoryAtURL:newBundleURL withIntermediateDirectories:FALSE attributes:nil error:nil];
+ 
+ 			return [[SQRLInstallerOwnedBundle alloc] initWithOriginalURL:request.targetBundleURL temporaryURL:newBundleURL codeSignature:codeSignature];
+ 		}]
+@@ -481,10 +482,50 @@ - (RACSignal *)installItemToURL:(NSURL *)targetURL fromURL:(NSURL *)sourceURL {
+ 	NSParameterAssert(targetURL != nil);
+ 	NSParameterAssert(sourceURL != nil);
+ 
++	NSLog(@"Moving bundle from %@ to %@", sourceURL, targetURL);
++
++	// If both the sourceURL and the targetURL exist we can try to skip a permissions check
++	// by moving Thing.app/Contents directly.  This allows us to update applications without
++	// permission to write files into the parent directory of Thing.app
++	//
++	// There is no known case where these directories don't exist but in order to handle
++	// edge cases / race conditions we'll handle it anyway.
++	//
++	// This exists check is non-atomic with the rename call below but that's OK
++	BOOL canRenameContentsDirectly = FALSE;
++	// For now while this is tested at scale this new option is behind a user default, this
++	// can be set by applications wishing to test this feature at runtime.  If it causes issues
++	// it can be opted out by individual users by setting this key to false explicitly.
++	// Once this has bene tested at scale it will become the default for all Squirrel.Mac
++	// users.
++	NSUserDefaults *defaults = [[NSUserDefaults alloc] init];
++	[defaults addSuiteNamed:_applicationIdentifier];
++	// In cases where this code is being executed under the ShipIt executable it's running
++	// under an application identifier equal to {parent_identifier}.ShipIt
++	// In this case we need to use the true parent identifier too as that is 99% of the time
++	// where the key will be set.
++	if ([_applicationIdentifier hasSuffix:@".ShipIt"]) {
++		[defaults addSuiteNamed:[_applicationIdentifier substringToIndex:[_applicationIdentifier length] - 7]];
++	}
++
++	if ([defaults boolForKey:@"SquirrelMacEnableDirectContentsWrite"]) {
++		canRenameContentsDirectly = [NSFileManager.defaultManager fileExistsAtPath:targetURL.path] && [NSFileManager.defaultManager fileExistsAtPath:sourceURL.path];
```

Comment:
Do we want to check permissions or anything?

Reply:
Permissions are checked by the calling process, that's what `launchPrivileged` is for.